app/models/product.rb

```diff
@@ -2,11 +2,14 @@ class Product < ActiveRecord::Base
 # VALIDATIONS ------------------------------------------------------------------
 
  validates :name, presence: true, uniqueness: true
- validates :price, presence: true, numericality: {only_float: true, greater_than: 0}
+ validates :price, presence: true, numericality: { greater_than: 0 }
+ validates :user_id, presence: true
+ validates :stock, numericality: {greater_than_or_equal_to: 0}
 
 # ASSOCIATIONS -----------------------------------------------------------------
 
   has_and_belongs_to_many :categories
+  accepts_nested_attributes_for :categories#, reject_if: proc {|attributes| attributes['name'].blank? }
   belongs_to :user
   has_many :reviews
   has_many :order_items
```

Comment:
I think the line below on #22 could be simplified to 

`self.retired = self.retired ? false : true`